### PR TITLE
INS-56: object has reference

### PR DESCRIPTION
**Pull request by ilyapt**
**- What I did**
- Implemented BaseCallable as a base implementation of Callable
- Implemented methods GetReference and SetReference
- Moved Callable and BaseCallable to callable.go
- BaseSmartContract, reference and factories inherit BaseCallable

**- How to verify it**
- Run tests

### CODE REVIEW

File: genesis/model/object/doc.go [hunk 1]
```diff
@@ -26,5 +26,13 @@ Usage:
 		ref, err := NewReference(domain, record, GlobalScope)
 	}
 
+
+Callable allows itself to be called by its reference
+BaseCallable is a base implementation of Callable
+
+Usage:
+    GetReference() - returns reference
+	SetReference(reference) - sets reference
```
Comment (LinnikD) [inline genesis/model/object/doc.go]: Need to format spaces

File: genesis/model/object/callable.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	SetReference(reference Reference)
+}
+
+// BaseCallable is a base implementation of Callable.
+type BaseCallable struct {
+	BaseObject
+	reference Reference
+}
+
+// GetReference return reference.
```
Comment (megge-dream) [inline genesis/model/object/callable.go]: GetReference return**s** ...

File: genesis/model/object/doc.go (hunk 1 above)
Comment (megge-dream) [inline genesis/model/object/doc.go]: Align lines

File: genesis/model/object/reference.go [hunk 3]
```diff
@@ -86,13 +87,6 @@ func (r *reference) String() string {
 	return fmt.Sprintf("#%s.#%s", r.domain, r.record)
 }
 
-// GetReference return reference to Reference instance.
-// For now Reference is Proxy and have itself as its reference.
-// TODO: ProxyReference & ReferenceInterface
```
Comment (megge-dream) [inline genesis/model/object/reference.go]: Save this comment for `type reference struct`

File: genesis/model/resolver/global.go [hunk 4]
```diff
@@ -57,6 +57,7 @@ func (r *globalResolver) GetObject(reference interface{}, cls interface{}) (inte
 	if proxy.GetClassID() != classID {
 		return nil, fmt.Errorf("instance class is not `%s`", classID)
 	}
+	proxy.SetReference(reference.(object.Reference))
```
Comment (megge-dream) [inline genesis/model/resolver/global.go]: Here `ref, ok := reference.(object.Reference)` you already got reference as object.Reference, so you can do just `proxy.SetReference(ref)` now

File: genesis/model/resolver/handler_test.go [hunk 5]
```diff
@@ -91,9 +94,11 @@ func TestHandler_GetObject_GlobalScope(t *testing.T) {
 	(*GlobalResolver.globalInstanceMap)["123"] = mockParent
 
 	obj, err := resolverHandler.GetObject(ref, "mockChild")
+	rf := obj.(*mockChild).GetReference()
```
Comment (megge-dream) [inline genesis/model/resolver/handler_test.go]: `obj.(object.Callable).GetReference()` is better - you bind to common interface, not to specific struct

File: genesis/model/resolver/proxy.go [hunk 6]
```diff
@@ -35,6 +35,11 @@ func (bp *BaseProxy) GetReference() object.Reference {
 	return bp.Instance.GetReference()
 }
 
+// SetReference proxy call for instance method.
```
Comment (megge-dream) [inline genesis/model/resolver/proxy.go]: SetReference **is a** proxy ...

File: genesis/model/object/doc.go (hunk 1 above)
Comment (LinnikD) [inline genesis/model/object/doc.go]: format spaces

File: genesis/model/resolver/proxy_test.go [hunk 7]
```diff
@@ -44,11 +50,13 @@ func TestBaseProxy_GetClassID(t *testing.T) {
 	assert.Equal(t, "mockChild", proxy.GetClassID())
 }
 
-func TestBaseProxy_GetReference(t *testing.T) {
+func TestBaseProxy_SetGetReference(t *testing.T) {
```
Comment (megge-dream) [inline genesis/model/resolver/proxy_test.go]: It should be two different tests

File: genesis/public/core/class.go [hunk 8]
```diff
@@ -140,6 +141,9 @@ func (cdf *classDomainFactory) GetReference() object.Reference {
 	return nil
 }
 
+func (cdf *classDomainFactory) SetReference(ref object.Reference) {
```
Comment (megge-dream) [inline genesis/public/core/class.go]: Why is it empty?

File: genesis/public/core/reference.go [hunk 9]
```diff
@@ -162,7 +163,10 @@ func (rdf *referenceDomainFactory) GetReference() object.Reference {
 	return nil
 }
 
-// Create is a factory method for new ReferenceDomain instances.
+func (rdf *referenceDomainFactory) SetReference(reference object.Reference) {
```
Comment (LinnikD) [inline genesis/public/core/reference.go]: Add doc string

File: genesis/public/core/instance.go [hunk 10]
```diff
@@ -147,6 +148,10 @@ func (idf *instanceDomainFactory) GetReference() object.Reference {
 	return nil
 }
 
+func (idf *instanceDomainFactory) SetReference(reference object.Reference) {
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: Why is it empty?

File: genesis/public/core/reference.go (hunk 9 above)
Comment (megge-dream) [inline genesis/public/core/reference.go]: Comment for SetReference

File: genesis/public/core/class.go (hunk 8 above)
Comment (LinnikD) [reply genesis/public/core/class.go]: This method should be removed

File: genesis/example/member_test.go [hunk 11]
```diff
@@ -107,11 +111,16 @@ func (p *mockProxy) GetReference() object.Reference {
 	return nil
 }
 
+func (p *mockProxy) SetReference(reference object.Reference) {
```
Comment (megge-dream) [inline genesis/example/member_test.go]: Why not inherit mockProxy from mockCallable instead (as with mockFactory, mockParent and mockChild)?

File: genesis/model/object/reference.go [hunk 12]
```diff
@@ -89,9 +90,6 @@ func (r *reference) String() string {
 // GetReference return reference to Reference instance.
 // For now Reference is Proxy and have itself as its reference.
 // TODO: ProxyReference & ReferenceInterface
```
Comment (megge-dream) [inline genesis/model/object/reference.go]: This comment should be at line 44, not in the middle of methods.

Comment (megge-dream) [reply genesis/model/object/reference.go]: Add without "GetReference return reference to Reference instance" line

File: genesis/model/resolver/global.go [hunk 13]
```diff
@@ -57,6 +57,7 @@ func (r *globalResolver) GetObject(reference interface{}, cls interface{}) (inte
 	if proxy.GetClassID() != classID {
 		return nil, fmt.Errorf("instance class is not `%s`", classID)
 	}
+	proxy.SetReference(ref)
```
Comment (LinnikD) [inline genesis/model/resolver/global.go]: You should do it in every resolver implementation, not only global

File: genesis/public/core/reference.go [hunk 14]
```diff
@@ -157,12 +158,7 @@ func (rdf *referenceDomainFactory) GetClassID() string {
 	return class.ReferenceDomainID
 }
 
-// GetReference returns nil for not published factory.
-func (rdf *referenceDomainFactory) GetReference() object.Reference {
-	return nil
-}
-
-// Create is a factory method for new ReferenceDomain instances.
+// Create factory method for new ReferenceDomain instances.
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: Create **is a** factory method ...

**Pull request merged**